File: src/history.ts
import * as path from "@std/path";
import { exists } from "@std/fs";

export interface HistoryEntry {
  timestamp: string;
  requestFile: string;
  requestName?: string;
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: string;
  responseStatus: number;
  responseStatusText: string;
  responseHeaders: Record<string, string>;
  responseBody: string;
  duration: number;
  error?: string;
}

export class HistoryManager {
  private historyDir: string;

  constructor(baseDir?: string) {
    // If baseDir not provided, use current directory for backward compatibility
    const dir = baseDir ?? ".";
    this.historyDir = path.join(dir, "history");
  }

  /**
   * Initialize history directory
   */
  async init(): Promise<void> {
    if (!await exists(this.historyDir)) {
      await Deno.mkdir(this.historyDir, { recursive: true });
    }
  }

  /**
   * Generate timestamp for filename (ISO 8601 format, safe for filenames)
   */
  private generateTimestamp(): string {
    const now = new Date();
    return now.toISOString()
      .replace(/:/g, "-")  // Replace colons for filename safety
      .replace(/\./g, "-") // Replace dots for filename safety
      .slice(0, -1);       // Remove trailing 'Z'
  }

  /**
   * Save a history entry to file
   */
  async save(entry: HistoryEntry): Promise<string> {
    await this.init();

    const filenameSafeTimestamp = this.generateTimestamp();

    // Extract base filename from request file path
    const requestBaseName = path.basename(entry.requestFile, ".http");
    const historyFileName = `${requestBaseName}_${filenameSafeTimestamp}.json`;
    const historyFilePath = path.join(this.historyDir, historyFileName);

    // Save entry with original ISO timestamp (keep it parseable)
    await Deno.writeTextFile(
      historyFilePath,
      JSON.stringify(entry, null, 2)
    );

    return historyFilePath;
  }

  /**
   * Get all history entries for a specific request file
   */
  async getHistory(requestFile: string): Promise<HistoryEntry[]> {
    if (!await exists(this.historyDir)) {
      return [];
    }

    const requestBaseName = path.basename(requestFile, ".http");
    const entries: HistoryEntry[] = [];

    for await (const dirEntry of Deno.readDir(this.historyDir)) {
      if (dirEntry.isFile && dirEntry.name.startsWith(requestBaseName) && dirEntry.name.endsWith(".json")) {
        const filePath = path.join(this.historyDir, dirEntry.name);
        const content = await Deno.readTextFile(filePath);
        const entry: HistoryEntry = JSON.parse(content);
        entries.push(entry);
      }
    }

    // Sort by timestamp (newest first)
    entries.sort((a, b) => b.timestamp.localeCompare(a.timestamp));

    return entries;
  }

  /**
   * Clear history for a specific request file
   */
  async clearHistory(requestFile: string): Promise<number> {
    if (!await exists(this.historyDir)) {
      return 0;
    }

    const requestBaseName = path.basename(requestFile, ".http");
    let count = 0;

    for await (const dirEntry of Deno.readDir(this.historyDir)) {
      if (dirEntry.isFile && dirEntry.name.startsWith(requestBaseName) && dirEntry.name.endsWith(".json")) {
        const filePath = path.join(this.historyDir, dirEntry.name);
        await Deno.remove(filePath);
        count++;
      }
    }

    return count;
  }

  /**
   * Clear all history
   */
  async clearAllHistory(): Promise<number> {
    if (!await exists(this.historyDir)) {
      return 0;
    }

    let count = 0;
    for await (const dirEntry of Deno.readDir(this.historyDir)) {
      if (dirEntry.isFile && dirEntry.name.endsWith(".json")) {
        const filePath = path.join(this.historyDir, dirEntry.name);
        await Deno.remove(filePath);
        count++;
      }
    }

    return count;
  }

  /**
   * Get the latest history entry for a request file
   */
  async getLatest(requestFile: string): Promise<HistoryEntry | null> {
    const history = await this.getHistory(requestFile);
    return history.length > 0 ? history[0] : null;
  }
}

File: src/internal/history/history.go
package history

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/studiowebux/restcli/internal/config"
	"github.com/studiowebux/restcli/internal/types"
)

// Save saves a request/response pair to history
func Save(requestFile string, req *types.HttpRequest, result *types.RequestResult) error {
	// Create history entry
	entry := types.HistoryEntry{
		Timestamp:          time.Now().Format(time.RFC3339),
		RequestFile:        requestFile,
		RequestName:        req.Name,
		Method:             req.Method,
		URL:                req.URL,
		Headers:            req.Headers,
		Body:               req.Body,
		ResponseStatus:     result.Status,
		ResponseStatusText: result.StatusText,
		ResponseHeaders:    result.Headers,
		ResponseBody:       result.Body,
		Duration:           result.Duration,
		RequestSize:        result.RequestSize,
		ResponseSize:       result.ResponseSize,
		Error:              result.Error,
	}

	// Generate filename: {requestBaseName}_{timestamp}.json
	baseName := filepath.Base(requestFile)
	baseName = strings.TrimSuffix(baseName, filepath.Ext(baseName))
	timestamp := time.Now().Format("20060102_150405")
	filename := fmt.Sprintf("%s_%s.json", baseName, timestamp)

	historyPath := filepath.Join(config.HistoryDir, filename)

	// Marshal to JSON
	data, err := json.MarshalIndent(entry, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal history entry: %w", err)
	}

	// Write to file
	if err := os.WriteFile(historyPath, data, config.FilePermissions); err != nil {
		return fmt.Errorf("failed to write history file: %w", err)
	}

	return nil
}

// Load loads all history entries
func Load() ([]types.HistoryEntry, error) {
	entries, err := os.ReadDir(config.HistoryDir)
	if err != nil {
		if os.IsNotExist(err) {
			return []types.HistoryEntry{}, nil
		}
		return nil, fmt.Errorf("failed to read history directory: %w", err)
	}

	var history []types.HistoryEntry

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
			continue
		}

		filePath := filepath.Join(config.HistoryDir, entry.Name())
		data, err := os.ReadFile(filePath)
		if err != nil {
			continue // Skip files that can't be read
		}

		var histEntry types.HistoryEntry
		if err := json.Unmarshal(data, &histEntry); err != nil {
			continue // Skip files that can't be parsed
		}

		history = append(history, histEntry)
	}

	// Sort by timestamp (newest first)
	sort.Slice(history, func(i, j int) bool {
		return history[i].Timestamp > history[j].Timestamp
	})

	return history, nil
}

// LoadForFile loads history entries for a specific request file
func LoadForFile(requestFile string) ([]types.HistoryEntry, error) {
	baseName := filepath.Base(requestFile)
	baseName = strings.TrimSuffix(baseName, filepath.Ext(baseName))

	entries, err := os.ReadDir(config.HistoryDir)
	if err != nil {
		if os.IsNotExist(err) {
			return []types.HistoryEntry{}, nil
		}
		return nil, fmt.Errorf("failed to read history directory: %w", err)
	}

	var history []types.HistoryEntry

	// Pattern: {baseName}_*.json
	prefix := baseName + "_"

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasPrefix(entry.Name(), prefix) {
			continue
		}

		filePath := filepath.Join(config.HistoryDir, entry.Name())
		data, err := os.ReadFile(filePath)
		if err != nil {
			continue
		}

		var histEntry types.HistoryEntry
		if err := json.Unmarshal(data, &histEntry); err != nil {
			continue
		}

		history = append(history, histEntry)
	}

	// Sort by timestamp (newest first)
	sort.Slice(history, func(i, j int) bool {
		return history[i].Timestamp > history[j].Timestamp
	})

	return history, nil
}

// Clear deletes all history entries
func Clear() error {
	entries, err := os.ReadDir(config.HistoryDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("failed to read history directory: %w", err)
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		filePath := filepath.Join(config.HistoryDir, entry.Name())
		if err := os.Remove(filePath); err != nil {
			// Continue even if some files fail to delete
			continue
		}
	}

	return nil
}

// Delete deletes a specific history entry by filename
func Delete(filename string) error {
	filePath := filepath.Join(config.HistoryDir, filename)
	return os.Remove(filePath)
}

// GetCount returns the total number of history entries
func GetCount() (int, error) {
	entries, err := os.ReadDir(config.HistoryDir)
	if err != nil {
		if os.IsNotExist(err) {
			return 0, nil
		}
		return 0, fmt.Errorf("failed to read history directory: %w", err)
	}

	count := 0
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".json") {
			count++
		}
	}

	return count, nil
}
